File: app/src/main/java/com/tamercad/ui/RenderViewport.kt
package com.tamercad.core

/**
 * Viewport sınırlarını tutan veri sınıfı.
 */
data class ViewportRectangle(
    var x: Float,
    var y: Float,
    var width: Float,
    var height: Float
)

class RenderViewport(
    width: Float = 800f,
    height: Float = 600f
) {
    private var rectangle = ViewportRectangle(0f, 0f, width, height)
    
    var pixelRatio: Float = 1.0f
        set(value) {
            require(value > 0) { "Pixel ratio must be greater than zero." }
            field = value
        }

    var isEnabled: Boolean = true
        private set

    fun setPosition(x: Float, y: Float) {
        rectangle.x = x
        rectangle.y = y
    }

    fun setSize(width: Float, height: Float) {
        rectangle.width = width
        rectangle.height = height
    }

    fun resize(width: Float, height: Float) {
        setSize(width, height)
    }

    fun getWidth(): Float = rectangle.width

    fun getHeight(): Float = rectangle.height

    fun getAspectRatio(): Float {
        return if (rectangle.height == 0f) {
            1f
        } else {
            rectangle.width / rectangle.height
        }
    }

    fun getRectangle(): ViewportRectangle {
        return rectangle.copy()
    }

    fun getPhysicalWidth(): Float {
        return rectangle.width * pixelRatio
    }

    fun getPhysicalHeight(): Float {
        return rectangle.height * pixelRatio
    }

    fun enable() {
        isEnabled = true
    }

    fun disable() {
        isEnabled = false
    }

    /**
     * Jetpack Compose Canvas (DrawScope) veya Android Canvas'a 
     * viewport ayarlarını uygulamak için kullanılacak fonksiyon.
     */
    fun applyViewport(context: Any?) {
        if (!isEnabled) return

        // İleride OpenGL ES veya Vulkan'a geçildiğinde (Mimari Faz 2/3),
        // JNI üzerinden C++ tarafına veya Android GLES30.glViewport()
        // metoduna bu değerler aktarılacak.
        
        /* Örnek OpenGL ES kullanımı:
         * GLES30.glViewport(
         *     rectangle.x.toInt(),
         *     rectangle.y.toInt(),
         *     getPhysicalWidth().toInt(),
         *     getPhysicalHeight().toInt()
         * )
         */
    }

    fun clone(): RenderViewport {
        val viewport = RenderViewport(rectangle.width, rectangle.height)
        viewport.setPosition(rectangle.x, rectangle.y)
        viewport.pixelRatio = this.pixelRatio
        if (!this.isEnabled) {
            viewport.disable()
        }
        return viewport
    }
}

File: RenderViewport.ts
export interface ViewportRectangle {

    x: number;

    y: number;

    width: number;

    height: number;

}


export class RenderViewport {

    private rectangle: ViewportRectangle = {

        x: 0,

        y: 0,

        width: 800,

        height: 600

    };


    private pixelRatio = 1.0;


    private enabled = true;


    constructor(

        width = 800,

        height = 600

    ) {

        this.setSize(

            width,

            height

        );

    }



    setPosition(

        x: number,

        y: number

    ): void {

        this.rectangle.x = x;

        this.rectangle.y = y;

    }



    setSize(

        width: number,

        height: number

    ): void {

        this.rectangle.width = width;

        this.rectangle.height = height;

    }



    resize(

        width: number,

        height: number

    ): void {

        this.setSize(

            width,

            height

        );

    }



    getWidth(): number {

        return this.rectangle.width;

    }



    getHeight(): number {

        return this.rectangle.height;

    }



    getAspectRatio(): number {

        if (

            this.rectangle.height === 0

        ) {

            return 1;

        }


        return (

            this.rectangle.width /

            this.rectangle.height

        );

    }



    getRectangle():

    ViewportRectangle {

        return {

            ...this.rectangle

        };

    }



    setPixelRatio(

        ratio: number

    ): void {

        if (

            ratio <= 0

        ) {

            throw new Error(

                "Pixel ratio must be greater than zero."

            );

        }


        this.pixelRatio = ratio;

    }



    getPixelRatio(): number {

        return this.pixelRatio;

    }



    getPhysicalWidth(): number {

        return (

            this.rectangle.width *

            this.pixelRatio

        );

    }



    getPhysicalHeight(): number {

        return (

            this.rectangle.height *

            this.pixelRatio

        );

    }



    enable(): void {

        this.enabled = true;

    }



    disable(): void {

        this.enabled = false;

    }



    isEnabled(): boolean {

        return this.enabled;

    }



    apply(

        context: any

    ): void {

        /**
         * GPU viewport uygulaması
         *
         * WebGL:
         *
         * gl.viewport(
         *    x,
         *    y,
         *    width,
         *    height
         * )
         *
         */


        if (

            !this.enabled

        ) {

            return;

        }



        if (

            context &&

            typeof context.viewport ===

            "function"

        ) {


            context.viewport(

                this.rectangle.x,

                this.rectangle.y,

                this.getPhysicalWidth(),

                this.getPhysicalHeight()

            );

        }

    }



    clone(): RenderViewport {

        const viewport =

            new RenderViewport(

                this.rectangle.width,

                this.rectangle.height

            );


        viewport.setPosition(

            this.rectangle.x,

            this.rectangle.y

        );


        viewport.setPixelRatio(

            this.pixelRatio

        );


        viewport.enabled =

            this.enabled;


        return viewport;

    }



    toJSON() {

        return {

            rectangle:

                this.rectangle,

            pixelRatio:

                this.pixelRatio,

            enabled:

                this.enabled

        };

    }



    static fromJSON(

        data: any

    ): RenderViewport {


        const viewport =

            new RenderViewport(

                data.rectangle.width,

                data.rectangle.height

            );


        viewport.setPosition(

            data.rectangle.x,

            data.rectangle.y

        );


        viewport.setPixelRatio(

            data.pixelRatio

        );


        if (

            !data.enabled

        ) {

            viewport.disable();

        }


        return viewport;

    }

}